helix-rest/src/main/java/org/apache/helix/rest/server/ServerContext.java

```diff
@@ -165,5 +238,93 @@ public void close() {
     if (_zkMetadataStoreDirectory != null) {
       _zkMetadataStoreDirectory.close();
     }
+    if (_zkClientForListener != null) {
+      _zkClientForListener.close();
+    }
+  }
+
+  @Override
+  public void handleChildChange(String parentPath, List<String> currentChilds) {
+    if (_zkClientForListener == null || _zkClientForListener.isClosed()) {
+      return;
+    }
+    // Resubscribe
+    _zkClientForListener.unsubscribeAll();
+    ZkRoutingDataReader.subscribeRoutingDataChanges(_zkClientForListener, this, this);
+    resetZkResources();
+  }
+
+  @Override
+  public void handleDataChange(String dataPath, Object data) {
+    if (_zkClientForListener == null || _zkClientForListener.isClosed()) {
+      return;
+    }
+    resetZkResources();
+  }
+
+  @Override
+  public void handleDataDeleted(String dataPath) {
+    if (_zkClientForListener == null || _zkClientForListener.isClosed()) {
+      return;
+    }
+    // Resubscribe
+    _zkClientForListener.unsubscribeAll();
+    ZkRoutingDataReader.subscribeRoutingDataChanges(_zkClientForListener, this, this);
+    resetZkResources();
```

Comment:
Is it possible to reduce duplication among the handle functions?

Reply:
Updated. It's hard to reduce duplication further because we need to sneak in unsubscribe/resubscribe for some callback functions.